### ISSUE TITLE
Enhanced Damage/Defense

### ISSUE THREAD

**Issue by scythetleppo** [hectorgimenez/d2go#15]
Hello,
Looking through d2go code; it looks like you're getting the item stats directly from memory. Can you confirm, then, that meta stats such as Enhanced Defense, Enhanced Damage do not work at all for d2go and koolo?

I've been putting lines like this in the NIP files:
[name] == sharkskinbelt && [quality] == unique && [flag] != ethereal # [enhanceddefense] >= 150

but now I'm realizing this will never be true, as there is never a value for enhanceddefense.

Can you confirm that please? I would like to change my NIPs if that's the case.

**Comment by hectorgimenez** [comment]
Yes, you are right, are not working. Are those stats calculated somehow?

**Comment by hectorgimenez** [comment]
For an Spirit Shield, this is what we read:

![image](https://github.com/hectorgimenez/d2go/assets/12543131/e70ca5eb-3ae3-466b-ac79-902c52fd633d)

And this is the object:
![image](https://github.com/hectorgimenez/d2go/assets/12543131/fbef1b36-3bf7-4af4-92b3-9e41a68223a7)

**Comment by scythetleppo** [comment]
Well, I am not an expert in memory, but I do have a working python memory reader that gets every piece of information from D2R that I know how to get. In a sense, my [private] project is similar to d2go (but is in python). I could be wrong about the following, so you should try to consult with someone else if possible. But as far as I understand it, there are only two ways to get these values:

1. The one I am not currently able to do (100% accuracy, fast)
2. The one I am probably able to do (90+% accuracy, slow)
3. The one I am currently doing (varying accuracy)

**The problem**
Let me first explain the issue in short (again, my understanding, could be incorrect):

The game doesn't care what the enhanced defense or enhanced damage or resist all is. These values are only for item tool tips. The game only cares what _defense_ or _damage_ or _each resist_ the item has. This makes sense when you think about it, because those are the things actually used in calculations. So these three are actually not item stats - they are tool tip stats. They are used on the server side for generating the item, but are only sent to the client for tool tips.

**1. The solution that would work 100% of the time, and is fast, that I don't know how to do**
If you roll your mouse over an item and scan memory for the ED value as string, for example **+15% Enhanced Damage**, you will find the tool tip lines represented pretty clearly. However, I'm not aware of how to reliably find these tool tips (the offsets), and they do seem to be fleeting (the value seems to move around and change). The best solution I know of would be to read the tool tip directly from in-game memory. Unfortunately I'm not good enough to do this myself.

**2. The one I am probably able to do (90+% accuracy, slow)**
It should be possible, knowing about the above, to scan the memory in real time and determine the tool tip being shown and get the strings. Basically, you need to figure out the memory offset range from the base address, that these tool tips are always showing in. Then you need to know something about the item you are rolling the mouse over. For example you know you are hovering an identified Shako. Then you scan in reverse order (as the tool tips lines are listed backwards), to find the Shako lines, then read the rest to get the values. This is potentially error-prone, because like I said the values seem to pop in and out of existence, so you may need to scan in a loop until you get all the values. I will probably work on this method myself soon.

**3.  The one I am currently doing (varying accuracy)**
The only way I know of to avoid any new memory scanning or reading, is to apply the following logic (pseudo-code). Though it's not pictured in your screenshot, you should be able to access the item's _base stats_ in addition to its _stats_. I will show you the enhanced **defense** logic, because it is more clear. The damage logic should be fairly obvious after reading this.
````
# this is not the entire code. you also need to calculate for the base item of runewords, etc. This is just an example so you get the basic idea.

defense = defense in "stats"
for item in the item's sockets:
  defense -= any defense from those items

if item.quality is unique or item.quality is set or item.is_runeword:
 properties = get the properties from properties.txt
 does it only add enhanced defense (no +defense) ?
 yes: is it always the same amount of ED?
     yes: you have your answer
     no: 
        defense -= any static defense added if it is always the same (for example magefist always adds 15 defense)
        enhanced defense = the percent increase from base defense to defense
        (if the item adds varying defense and varying enhanced defense, there is no way to know accurately which one is which)
````
There is more to it but that's an example of the logic needed. It is very accurate for some items (for example Fortitude Runeword armor. It has static +15 defense and 300% ed, so all you need to do is calculate the 0-15% superior item defense). It is extremely _inaccurate_ for items with small values (for example magefist, because it only has 24-25 defense, but has a 10-range enhanced defense). It is also inaccurate for items which add varying +defense and varying enhanced defense at the same time. So it is not a great solution, which is why I may move to scanning memory range soon, if it's not too slow.

For your spirit shield, since the runeword doesn't add any defense, the calculation is just 
````round_up ((170 - 148)/148) * 100 = 15%````
